pkg/config/validator.go

```diff
@@ -421,9 +550,83 @@ func (p *validator) validate(cfg map[string]string) (rt *Validated, ce *adapter.
 			return rt, ce.Appendf("serviceConfig", "failed validation").Extend(re)
 		}
 	}
+
+	for _, kk := range keymap[constructors] {
+		ck := parseRulesKey(kk)
+		if ck == nil {
+			continue
+		}
+		if re := p.validateConstructorConfigs(cfg[kk]); re != nil {
+			return rt, ce.Appendf("serviceConfig", "failed validation").Extend(re)
+		}
+	}
+
+	for _, kk := range keymap[actionRules] {
+		ck := parseRulesKey(kk)
+		if ck == nil {
+			continue
+		}
+		if re := p.validateRulesConfig(cfg[kk]); re != nil {
+			return rt, ce.Appendf("serviceConfig", "failed validation").Extend(re)
+		}
+	}
+
+	// everything is validated we can configure the handlers.
+	if re := p.buildHandlers(); re != nil {
+		return rt, ce.Extend(re)
+	}
+
 	return p.validated, nil
 }
 
+// TODO: What about partial failures. In the middle of building handlers,
```

Comment:
We promise to call close on everything we start, so for partial failures we should loop, create as many as possible, and if we fail any `Close` them all after we aggregate as many errors as possible.